csrc/device_lower/pass/circular_buffer.cpp

```diff
@@ -215,6 +216,706 @@ class CircularBufferLoopCloner : public kir::IrVisitor {
+// - All threads wait until tma operation arrives
+// - Copy body without
+//   - shared memory allocations
+//   - issuing cp async bulk operations
+//   - mbarrier_init exprs
+//   - mbarrier_inval exprs
+//
+// Post-epilogue loop:
+//  - if selected_thread:
+//   - Invalidated mbarrier for all stages
+//
+// Detailed Pseudo-Code:
+// Pre-Prologue loop:
+// __shared__ __mbarrier_t barriers[num_stages];
+// __shared__ __mbarrier_token_t tokens[num_stages];
+// if (warp_id == 0 && electSync()()) {
+//   for (int64_t loop_index : irange(stages)) {
+//     mbarrier_init(mbarrier[loop_index], number_of_arrival_threads);
+//   }
+// }
+//
+// Prologue loop:
+// for (int64_t loop_index : irange(stages-1)) {
+//   if (warp_id == 0 && electSync()()) {
+//     tokens[loop_index] =
+//       mbarrier::arriveExpectTx(mbarrier[loop_index], expected_bytes);
+//     for (...) {
+//       cpAsyncBulk(mbarriers[loop_index], ...);
+//     }
+//   } else {
+//     tokens[loop_index] =
+//       mbarrier::arrive(mbarrier[loop_index]);
+//   }
+// }
+//
+// Main loop:
+// for (int64_t loop_index : irange(N-(stages-1))) {
+//   current_stage = loop_index % stage_depth
+//   load_stage = (loop_index + (stage_depth - 1)) % stage_depth)
+//   if (warp_id == 0 && electSync()()) {
+//     token[load_stage] =
+//       mbarrier::arriveExpectTx(mbarrier[load_stage], expected_bytes);
+//     for (...) {
+//       cpAsyncBulk(mbarrier[load_stage], ...);
+//     }
+//   } else {
+//     token[load_stage] =
+//       mbarrier::arrive(mbarrier[load_stage]);
+//   }
+//   mbarrier::wait(token[current_stage]);
+//
+//   Clone remaining operations
+// }
+//
+// Epilogue loop:
+// for (int64_t loop_index : irange(N-(stages-1), N)) {
+//   current_stage = loop_index % stage_depth
+//   mbarrier::wait(token[current_stage]);
+//
+//   Clone remaining operations
+// }
+//
+// Post-Epilogue loop:
+// if (warp_id == 0 && electSync()()) {
+//   for (int64_t loop_index : irange(stages)) {
+//     mbarrier_inval(mbarrier[loop_index]);
+//   }
+// }
+//
+class TmaCircularBufferLoopCloner : public CircularBufferLoopCloner {
+ public:
+  static ForLoop* clone(
+      ForLoop* circular_buffer_loop,
+      const std::vector<Expr*>& circular_buffer_load_exprs,
+      CircularBufferLoopStage loop_type,
+      const std::unordered_set<Expr*>& exclude = {}) {
+    TmaCircularBufferLoopCloner cloner(
+        circular_buffer_loop, circular_buffer_load_exprs, loop_type, exclude);
+    cloner.duplicate();
+    return cloner.cloned_top_level_loop_;
+  }
+
+ private:
+  TmaCircularBufferLoopCloner(
+      ForLoop* circular_buffer_loop,
+      const std::vector<Expr*>& circular_buffer_load_exprs,
+      CircularBufferLoopStage loop_type,
+      const std::unordered_set<Expr*>& exclude)
+      : CircularBufferLoopCloner(
+            circular_buffer_loop,
+            circular_buffer_load_exprs,
+            loop_type,
+            exclude) {}
+
+  void processForLoop(ForLoop* cloned_loop) final {
```

Comment:
Could you add some comment about what this function does? The code here is far less trivial than `CircularBufferLoopCloner::processForLoop`, and I am having trouble understanding why.

Reply:
TL;DR: We have an mbarrier for each circular buffer stage and TensorView but not each TMA load, so we cannot add the `cpAsyncBulk` expressions sequentially.

```cpp
  // For CircularBufferLoopCloner, all load operations are synchronous, so they
  // are added sequentially. For TmaCircularBufferLoop, we have an mbarrier for
  // each Tensorview and each circular buffer stage, but not for each individual
  // TMA load operation. If there are serial IterDomains to the right of the
  // computeAt position, nvfuser will generate a for-loop to launch multiple TMA
  // load operations.
  //
  // When we encounter a CpAsyncBulk load expression, we create a mbarrier_wait
  // for the main and epilogue loops and a arriveExpectTx for prologue and main
  // loops. The expected_tx for arriveExpectTx is the cumulative transaction
  // size for all TMA load operations for the TensorView. Next, we generate the
  // nested for-loops for the serial IterDomains, but do not add them to the
  // cloned circular buffer loop immediately. Once the cloned circular buffer
  // loop is the only loop in the stack, add the arriveExpectTx and arrive
  // expressions, then the nested for-loop structure calling the TMA load
  // operations, and finally the mbarrier_wait.
```